addons/mass_mailing/models/mail_notification.py

```diff
@@ -0,0 +1,124 @@
+             'the related mail_mail can be deleted separately from its statistics. '
+             'However the ID is needed for several action and controllers.',
+        index=True)
+    email = fields.Char(string="Email")
+    message_id = fields.Char(string='Message-ID', help="Technical field for the email Message-ID (RFC 2392)")
+    # document
+    model = fields.Char(string='Document model')
+    res_id = fields.Many2oneReference(string='Document ID', model_field='model')
+    # campaign / wave data
+    # if mass_mailing_id is set the notification will act as a trace
+    mass_mailing_id = fields.Many2one('mailing.mailing', string='Mailing', index=True, ondelete='cascade')
+    campaign_id = fields.Many2one('utm.campaign', related='mass_mailing_id.campaign_id', string='Campaign', index=True, store=True)
+    # Bounce and tracking
+    canceled = fields.Datetime(help='Date when the email has been invalidated. '
+                                    'Invalid emails are blacklisted, opted-out or invalid email format')
+    scheduled = fields.Datetime(help='Date when the email has been created', default=fields.Datetime.now)
+    sent = fields.Datetime(help='Date when the email has been sent')
+    exception = fields.Datetime(help='Date of technical error leading to the email not being sent')
+    opened = fields.Datetime(help='Date when the email has been opened the first time')
+    replied = fields.Datetime(help='Date when this email has been replied for the first time.')
+    bounced = fields.Datetime(help='Date when this email has bounced.')
+    # Link tracking
+    links_click_ids = fields.One2many('link.tracker.click', 'mailing_trace_id', string='Links click')
+    clicked = fields.Datetime(help='Date when customer clicked on at least one tracked link')
+    # Status
+    notification_status = fields.Selection(selection_add=[
+        ('outgoing', 'Outgoing'),
+        ('opened', 'Opened'),
+        ('replied', 'Replied')])
+    status_update = fields.Datetime('State Update', help='Last state update of the mail')
+
+    def name_get(self):
+        return [(n.id, '%s: %s (%s)' % (n.notification_type, n.mass_mailing_id.name or n.res_partner_id.display_name or '', n.id)) for n in self]
+
+    def _update_trace_status(self, values):
+        status_whitelist = ['sent', 'opened', 'clicked', 'replied', 'bounced', 'exception', 'canceled']
+        if any(key in status_whitelist for key in values.keys()):
+            notification_status = next((key for key in values.keys() if key in status_whitelist), False)
+            if notification_status == 'clicked':
+                notification_status = 'opened'
+            return notification_status
+
+    @api.model_create_multi
+    def create(self, values_list):
+        for values in values_list:
+            # Change default values if mass mailing is set
+            if 'mass_mailing_id' in values and values.get('mass_mailing_id'):
```

Comment:
Equivalent to `if values.get('mass_mailing_id')`

(I have the strange feeling I already made this comment)